lib/galaxy/tools/xsd/galaxy.xsd

```diff
@@ -3752,6 +3752,18 @@ path in ``command`` and ``configfile`` elements. The name should not contain
 pipes or periods (e.g. ``.``).]]></xs:documentation>
       </xs:annotation>
     </xs:attribute>
+    <xs:attribute name="when_empty">
+      <xs:annotation>
+        <xs:documentation xml:lang="en"><![CDATA[Action to take when output
+is empty. Can currently be ``hide`` or ``delete``.]]></xs:documentation>
```

Comment:
Would be more accurate to say `when the output file size is 0.`

Reply:
You can also use the flag on collections to hide empty collections.

Reply:
> I think this will be difficult to predict in the scheduling of workflows and queueing up of tool runs. Say you have 2 matched collections with elements A B C, and you run them against each other, if one of the inputs got deleted we may either stop evaluating the workflow here or switch form matched to multiplied mode in the tool form, both of which is probably going to be a surprising event to the user.
> 
> This is not so much a concern if just hiding instead of deleting the output, but it still feels like something a user should have full control over.

I was just going to call it hide_when_empty as a boolean, but then maybe there are other things a tool designer wants to do when empty? First thought was delete. 

With a current tool I am working on, I am more concerned with allowing hiding (because of the reasons you listed, and others). But there could be a case where deleting is more desirable to a tool designer?

Reply:
> You can also use the flag on collections to hide empty collections.

Which is a different thing actually, this should probably be a separate attribute with a different name ? You can still have a collection with elements having no data when the attribute is set, which is different from how it works for normal datasets.

> But there could be a case where deleting is more desirable to a tool designer?

I hope not. It'll sound appealing to tool authors, but it will be a big pain when you try to do anything complex with such a tool. And as a user you will have no indication that this is going to happen.

Reply:
To expand a bit on the current use-case. There is a tool that creates between 1 and 110 (!, I know...) outputs (mix of datasets and collections; outputs are different file types as well), but which ones are going to be created cannot be readily determined before job run (so no prefiltering, unfortunately). 

For the datasets, they are using `from_work_dir`, for collections, it is using `discover_datasets`.

Reply:
> > You can also use the flag on collections to hide empty collections.
> 
> Which is a different thing actually, this should probably be a separate attribute with a different name ? You can still have a collection with elements having no data when the attribute is set, which is different from how it works for normal datasets.
> 

Don't think I agree that they are different. There is an OutputObject, it can have an action performed when_empy. For datasets this is file size, for collection this is content. 

Presumably, you might want to also apply this flag to the contents (datasets) within the collection, but that would happen to the dataset OutputObjects, not the collection OutputObject.


> > But there could be a case where deleting is more desirable to a tool designer?
> 
> I hope not. It'll sound appealing to tool authors, but it will be a big pain when you try to do anything complex with such a tool. And as a user you will have no indication that this is going to happen.

Perhaps, but I am not entirely convinced that this shouldn't be possible if a tool designer wants to do it. I suppose that I am -0 on removing delete functionality, but still think the `when_empty` attribute setup makes sense for potential future enhancements.

Reply:
@blankenberg Maybe the easiest solution is to fix the tool to not create the empty datasets in the first place? Or is there a use case for the empty datasets?

Reply:
I agree that not creating the datasets would be ideal, but unfortunately there is not a perfect way to determine the exact subset of files that will be created by the underlying binary for a particular set of inputs before job execution.

Another potential approach to handle this would be to use `discover_datasets` for all of the tool wrapper's outputs, but that would greatly decrease the utility of the tool, and, I would argue, be objectively worse than creating unused outputs (hidden or not). 

A third approach could be to fall back to a tool specific code file to perform the hiding, but I'd much rather avoid that. 


(Thanks to each of you for the reviews and discussions, it is appreciated)